extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3DataSegmentMover.java

```diff
@@ -193,4 +143,110 @@ public Void call() throws Exception
       throw Throwables.propagate(e);
     }
   }
+
+  /**
+   * Copies an object and after that checks that the object is present at the target location, via a separate API call.
+   * If it is not, an exception is thrown, and the object is not deleted at the old location. This "paranoic" check
+   * is added after it was observed that S3 may report a successful move, and the object is not found at the target
```

Comment:
Is there evidence this is actual data loss and not just eventual consistency?

Reply:
There is no evidence, but S3 docs prohibit this. However this change is all about protecting from S3 errors, not against S3 "specified behaviour". In this case, if we "move" a segment and it doesn't appear at the target location, who knows what does it mean.